examples/spring-hypermedia/src/test/java/io/wcm/caravan/rhyme/examples/spring/hypermedia/AbstractCompanyApiIntegrationTest.java

```diff
@@ -0,0 +1,250 @@
+import org.junit.jupiter.api.Test;
+import org.mockito.internal.util.collections.Iterables;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+
+import io.wcm.caravan.rhyme.api.exceptions.HalApiClientException;
+import io.wcm.caravan.rhyme.api.exceptions.HalApiServerException;
+
+/**
+ * Defines a set of tests that cover most of the API's functionality,
+ * but is only using the HAL API interfaces to navigate through resources.
+ */
+@SpringBootTest
+abstract class AbstractCompanyApiIntegrationTest {
+
+  private static final long NON_EXISTANT_ID = 999L;
+
+  @Autowired
+  private EmployeeRepository employeeRepository;
+  @Autowired
+  private ManagerRepository managerRepository;
+
+  private CompanyApi api;
+
+  @BeforeEach
+  void setUp() {
+    api = getApiImplementionOrClientProxy();
+  }
+
+  /**
+   * This will be overridden in the subclasses to return either the server-side implementation of
+   * the {@link CompanyApi} interface, or a dynamic client proxy that fetches the entry point
+   * with a HTTP request.
+   * @return the implementation of {@link CompanyApi} used to run the tests
+   */
+  protected abstract CompanyApi getApiImplementionOrClientProxy();
+
+  // The repositories are initialized with the same DatabaseLoader that
+  // is used when the application is started. For the tests that need to
+  // know an existing ID, we just pick the first one that exists in each repo.
+
+  private Long getIdOfFirstEmployee() {
+
+    return Iterables.firstOf(employeeRepository.findAll()).getId();
+  }
+
+  private Long getIdOfFirstManager() {
+
+    return Iterables.firstOf(managerRepository.findAll()).getId();
+  }
+
+  @Test
+  public void getEmployees_should_list_employees_in_order_of_creation() throws Exception {
+
+    List<EmployeeResource> employees = api.getEmployees().getAll();
+
+    assertThat(employees)
+        .extracting(employee -> employee.getState().getName())
+        .containsExactly("Frodo", "Bilbo", "Sam", "Pippin");
+  }
+
+  @Test
+  public void getManagers_should_list_managers_in_order_of_creation() throws Exception {
+
+    List<ManagerResource> managers = api.getManagers().getAll();
+
+    assertThat(managers)
+        .extracting(manager -> manager.getState().getName())
+        .containsExactly("Gandalf", "Saruman");
+  }
+
+  @Test
+  public void getEmployeeById_should_find_existing_employee() throws Exception {
+
+    Long firstId = getIdOfFirstEmployee();
+
+    Employee firstEmployee = api.getEmployeeById(firstId).getState();
+
+    assertThat(firstEmployee.getId()).isEqualTo(firstId);
+    assertThat(firstEmployee.getName()).isEqualTo("Frodo");
+    assertThat(firstEmployee.getRole()).isEqualTo("ring bearer");
+  }
+
+  @Test
+  public void getEmployeeById_should_respond_with_404_for_non_existing_id() throws Exception {
+
+    assertThat404isReturnedFor(
+        () -> api.getEmployeeById(NON_EXISTANT_ID).getState());
```

Comment:
One thing that is not intuitive here is that the exception with 404 status code is only thrown when #getState is called on the EmployeeResource returned by #getEmployeeId. This is especially true if this method is called directly on the CompanyApiController's implementation.

It would be nicer if #getEmployeeById would return an Optional<EmployeeResource> which would be empty if no employee exists for the given ID. However, using Optional<T> as a return type in a HalApiInterface already has a different meaning:. "the link (template) that corresponds to this method may not be present in every resource".

Also if #getEmployeeById is called on a client proxy, the proxy would have to actually fetch the resource from the resolved URI template immediately to figure out if the expanded link actually points to an existing resource. This would make the behaviour of client proxies a bit inconsistent, because in all other case it's guaranteed that a resource isn't fetched before a method on the corresponding proxy is being called.

Does anyone have an opinion what's more important here, or any other idea how to solve this?